pkg/util/strategicpatch/patch.go

```diff
@@ -0,0 +1,454 @@
+			// First find the fieldPatchStrategy and fieldPatchMergeKey.
+			fieldType, fieldPatchStrategy, fieldPatchMergeKey, err := lookupPatchMetadata(t, k)
+			if err != nil {
+				return nil, err
+			}
+
+			originalType := reflect.TypeOf(original[k])
+			patchType := reflect.TypeOf(patchV)
+			if originalType == patchType {
+				if originalType.Kind() == reflect.Map && fieldPatchStrategy != "replace" {
+					typedOriginal := original[k].(map[string]interface{})
+					typedPatch := patchV.(map[string]interface{})
+					var err error
+					original[k], err = mergeMap(typedOriginal, typedPatch, fieldType)
+					if err != nil {
+						return nil, err
+					}
+					continue
+				} else if originalType.Kind() == reflect.Slice && fieldPatchStrategy == "merge" {
+					elemType := fieldType.Elem()
+					typedOriginal := original[k].([]interface{})
+					typedPatch := patchV.([]interface{})
+					var err error
+					original[k], err = mergeSlice(typedOriginal, typedPatch, elemType, fieldPatchMergeKey)
+					if err != nil {
+						return nil, err
+					}
+					continue
+				}
+			}
+
+			original[k] = patchV
+		}
+	}
+
+	return original, nil
+}
+
+func mergeSlice(original, patch []interface{}, elemType reflect.Type, mergeKey string) ([]interface{}, error) {
+	if len(original) == 0 && len(patch) == 0 {
+		return original, nil
+	}
+	// All the values must be of the same type, but not a list.
+	t, err := sliceElementType(original, patch)
+	if err != nil {
+		return nil, fmt.Errorf("types of list elements need to be the same, type: %s: %v",
+			elemType.Kind().String(), err)
+	}
+	if t.Kind() == reflect.Slice {
+		return nil, fmt.Errorf("not supporting merging lists of lists yet")
+	}
+	// If the elements are not maps, merge the slices of scalars.
+	if t.Kind() != reflect.Map {
+		// Maybe in the future add a "concat" mode that doesn't
+		// uniqify.
+		both := append(original, patch...)
+		return uniqifyScalars(both), nil
+	}
+
+	if mergeKey == "" {
+		return nil, fmt.Errorf("cannot merge lists without merge key for type %s", elemType.Kind().String())
+	}
+
+	// First look for any special __patch elements.
+	patchWithoutSpecialElements := []interface{}{}
+	strict := false
+	for _, v := range patch {
+		typedV := v.(map[string]interface{})
+		patchType, ok := typedV[specialKey]
+		if ok {
+			if patchType == "delete" {
+				mergeValue, ok := typedV[mergeKey]
+				if ok {
+					_, originalKey, found := findMapInSliceBasedOnKeyValue(original, mergeKey, mergeValue)
+					if found {
+						// Delete the element at originalKey.
+						original = append(original[:originalKey], original[originalKey+1:]...)
+					}
+				} else {
+					return nil, fmt.Errorf("delete patch type with no merge key defined")
+				}
+			} else if patchType == "strict" {
+				strict = true
+				// Continue iterating through the array to prune any other __patch elements.
+			} else {
+				return nil, fmt.Errorf("unknown patch type found: %s", patchType)
```

Comment:
Users might expect to be able to specify `__patch: merge`, so we probably should generate a specific error message for that.